src/Psr17ResponseFactoryTrait.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mezzio\Hal;
+
+use Mezzio\Container\ResponseFactoryFactory;
+use Mezzio\Hal\Response\CallableResponseFactoryDecorator;
+use Psr\Container\ContainerInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @internal
+ * @deprecated Will be removed with v3.0.0
+ */
+trait Psr17ResponseFactoryTrait
+{
+    private function detectResponseFactory(ContainerInterface $container): ResponseFactoryInterface
+    {
+        $psr17FactoryAvailable = $container->has(ResponseFactoryInterface::class);
+
+        if (! $psr17FactoryAvailable) {
+            return $this->createResponseFactoryFromDeprecatedCallable($container);
+        }
+
+        if ($this->doesConfigurationProvidesDedicatedResponseFactory($container)) {
+            return $this->createResponseFactoryFromDeprecatedCallable($container);
+        }
+
+        $responseFactory = $container->get(ResponseFactoryInterface::class);
+        Assert::isInstanceOf($responseFactory, ResponseFactoryInterface::class);
+        return $responseFactory;
+    }
+
+    private function createResponseFactoryFromDeprecatedCallable(
+        ContainerInterface $container
+    ): ResponseFactoryInterface {
+        /** @var callable():ResponseInterface $responseFactory */
+        $responseFactory = $container->get(ResponseInterface::class);
+
+        return new CallableResponseFactoryDecorator($responseFactory);
+    }
+
+    private function doesConfigurationProvidesDedicatedResponseFactory(ContainerInterface $container): bool
```

Comment:
```suggestion
    private function configurationProvidesDedicatedResponseFactory(ContainerInterface $container): bool
```

Reads with a bit more flow.